[all] Add LDAP/STLP instructions

RaghavRoy145 opened the pull request:
Adds support for LDAP (Load-Acquire Pair) and STLP (Store-Release Pair) instructions to herd.

#### Summary

LDAP Xt1, Xt2, [Xn]:  Load-Acquire Pair. Both loads take Acquire ordering, subsequent memory accesses on the same thread cannot be reordered before the pair.
STLP Xt1, Xt2, [Xn]: Store-Release Pair. The pair takes Release ordering, earlier memory accesses cannot be reordered after it.

The existing `pair_opt = Pa | PaN | PaI` is refactored into polymorphic variants split by direction:  `ld_pair_opt` gains \`PaA (LDAP) and `st_pair_opt` gains \`PaL (STLP), while the old \`PaI is split into \`PaIQ (load side) and \`PaIL (store side). Polymorphic variants let load-only and store-only tags be enforced at the instruction level on I_LDP / I_STP without duplicating constructors like `Pa` and `PaN`, which are used on both the store and load directions.

#### Test Added:
`Y006.litmus` is a message-passing test with STLP as writer and LDAP as reader:

If the Release on the STLP and the Acquire on the LDAP compose correctly, then observing the second element of the pair (1:X3=1) must imply observing the preceding store, so 1:X5=0 should be impossible.

relokin commented on herd/AArch64Sem.ml:
```diff
@@ -1879,6 +1879,34 @@ Arguments:
+        
+        (* Perform the writes *)
+        (write_reg_dest r1 v1 ii >>| write_reg_dest r2 v2 ii)
+        >>= M.ignore >>= B.next1T
```
Can you please reuse the module LoadPair instead?

relokin reviewed (commented):
Thanks for this! I've got two comments.

RaghavRoy145 commented on herd/AArch64Sem.ml:
```diff
@@ -1879,6 +1879,34 @@ Arguments:
+        (write_reg_dest r1 v1 ii >>| write_reg_dest r2 v2 ii)
+        >>= M.ignore >>= B.next1T
+
+      let stlp sz r1 r2 ra ii =
```
Thanks for the reviews! Yes, I have made this change.

Commit:
[all] Refactor pair_opt type

Split pair_opt to load and store types to
allow load and store specific annotations
to be added.

Commit:
[all] Add LDAP/STLP instructions to herd

relokin reviewed (commented):
I am not sure we will retain backwards compatibility, but I think diyone7 will need to support the new instructions. This means we need the annotations PaL -> STLP and PaA -> LDAP. Now I think we can either keep interpreting PaI as either LDIAPP or STILP or we can split that to PaIQ and PaIL. Not that right now I can't generate a test with STILP. Neither this works:
```
$> diyone7 -metadata false -arch AArch64 PodWW PaIL Rfe PodRR Fre
```
nor this:
```
$> diyone7 -metadata false -arch AArch64 PodWW PaI Rfe PodRR Fre
```
and indeed neither annotation is valid
```
$> diyone7 -arch AArch64 -show annotations
 P X XL XA XAL L Q A PaL PaA PaIQ PaN Pa
```

HadrienRenaud commented on herd/AArch64ASLSem.ml:
```diff
@@ -866,42 +866,62 @@ module Make (TopConf : AArch64Sig.Config) (V : Value.AArch64ASL) :
+        | `PaL -> "ldiappstilp/STLP_64_ldiappstilp.opn"
+      and acqrel = false
+      and offset = if pa = `PaIL then liti k else litbv 64 k in
+  Some
```
nit: wrong intendation.

RaghavRoy145 commented on lib/AArch64Base.ml:
```diff
@@ -1226,7 +1226,25 @@ let tr_simd_variant = function
 
 type temporal = TT | NT
-type pair_opt = Pa | PaN | PaI
+type pair_opt = [`Pa | `PaN | `PaIQ | `PaIL | `PaA | `PaL]
```
- I've kept `pair_opt` because having separate `Ld_Pair` and `St_Pair` caused assertion failures in both `add_lxm_atom` and `add_lxm_edge` as the pretty-printers for shared variants like Pa and PaN produce identical strings, but `compare_atomo` sees them as different constructors. 
- Using prefixes (LdPa/StPa) breaks backward compatibility with diyone7 and user-facing tests.
- The unified `Pair` at the was the only approach that preserved backward compatibility while avoiding these issues, and this needs the `pair_opt` type. The split into `ld_pair_opt/st_pair_opt` is preserved at the instruction level (I_LDP/I_STP) where direction is known, with `pair_opt_to_ld `and `pair_opt_to_st` conversion functions used in `emit_access` where the direction is relevant again.

RaghavRoy145 replied:
I have instead replaced the usage of this type with `[ld_pair_opt | st_pair_opt]`

RaghavRoy145 commented on lib/AArch64Base.ml:
```diff
@@ -1226,7 +1226,25 @@ let tr_simd_variant = function
 type temporal = TT | NT
-type pair_opt = Pa | PaN | PaI
+type pair_opt = [`Pa | `PaN | `PaIQ | `PaIL | `PaA | `PaL]
+type ld_pair_opt = [`Pa | `PaN | `PaIQ | `PaA]
+type st_pair_opt = [`Pa | `PaN | `PaIL | `PaL]
```
I used polymorphic variants so that `ld_pair_opt` and `st_pair_opt` could share common tags (`Pa, `PaN) while having separate type constraints on `I_LDP` vs `I_STP`.

With regular variants, the shared constructors (Pa, PaN) would need to be duplicated across both types (e.g. LdPa/StPa, LdPaN/StPaN), and every conversion and pattern match becomes more verbose. If regular variants are preferred, it's a straightforward change. Happy to switch if there's a preference.

ShaleXIONG replied:
From the `diy` side, I have no strong opinion if you want to _rework_ constructors `Pa` and `PaN`, which originally can be apply to both `read` and `write` event. For example if we introduce different new name and hence remove the polymorphic type implementation here.

RaghavRoy145 commented on gen/common/AArch64Arch_gen.ml:
```diff
@@ -477,13 +477,24 @@ let is_tthm fields =
      | Some Capability -> "c"
 
    let pp_pair_opt = function
```
Currently `atom_acc `uses the `Pair of pair_opt * pair_idx`. 

At the instruction level, the pretty-printing of `I_LDP/I_STP` doesn't go through `pp_pair_opt` and it uses the instruction printer directly. So there's no need for split `pp_ld_pair_opt/pp_st_pair_opt ` functions. 

If we were to remove the unified `pair_opt` and go back to `Ld_Pair/St_Pair` in `atom_acc`, we cpould split printers, but then we'd hit the `add_lxm_atom` and `add_lxm_edge` assertion failures described above since both would produce the same strings for shared variants like `Pa` and `PaN`.

ShaleXIONG replied:
This implementation is fine for me. 

However can you extend the function `applies_atom`, which checks the validity between annotation, `a` and memory event, `d` (of `R` and `W`). It should be around line `439` in your branch, in this file. This enable an earlier filter in building cycle, rather an error raise in the final compilation time.

RaghavRoy145 replied:
I have made the extension to add the direction based matching in `applies_atom`

Though I think the above change should be made as it is a stronger check, this does not mean I can get rid of the above type-conversion functions (`pair_opt_to_ld `and `pair_opt_to_st`) unfortunately, as:

`emit_ldp` expects a ld_pair_opt (and equivalently the emit store analogue), but `opt` as it comes from Pair() has the following type:
`| Pair of [ld_pair_opt | st_pair_opt] * pair_idx`

So it is the union of the two pair types which `emit_ldp` does not like

We could keep the type-conversion functions, as well as make the change @ShaleXIONG  wants me to in `applies_atom` 

With the changes to `applies_atom` I get this error for a wrong cycle (example courtesy of @ShaleXIONG ):

```
diyone7 -arch AArch64 PaL PosRW Pa Rfi -oneloc      
diyone7: Fatal error: Test CoRW1+pospalpa-rfipapal [PosRWPaLPa RfiPaPaL] failed:
annotation mismatch on edge PosRWPaLPa, annotation 'PaL' on R
```

RaghavRoy145 commented:
> I am not sure we will retain backwards compatibility, but I think diyone7 will need to support the new instructions. This means we need the annotations PaL -> STLP and PaA -> LDAP. Now I think we can either keep interpreting PaI as either LDIAPP or STILP or we can split that to PaIQ and PaIL. Not that right now I can't generate a test with STILP. Neither this works:
> 
> ```
> $> diyone7 -metadata false -arch AArch64 PodWW PaIL Rfe PodRR Fre
> ```
> 
> nor this:
> 
> ```
> $> diyone7 -metadata false -arch AArch64 PodWW PaI Rfe PodRR Fre
> ```
> 
> and indeed neither annotation is valid
> 
> ```
> $> diyone7 -arch AArch64 -show annotations
>  P X XL XA XAL L Q A PaL PaA PaIQ PaN Pa
> ```


This I think is because of a bug in my PR, I failed to include "PaIL" in `fold_pair` (I overwrote it in [the second commit](https://github.com/herd/herdtools7/pull/1639/changes/cb76a56ec27d8d7ec4dc0ee6c257a3fd31d3d338#diff-9dda1bfaa29914247a91cd78d17def95cfca44cdbc1b45e52ee2e264175be1c3L596))

Here is a working example after I fixed it locally:

```
herdtools7 % _build/install/default/bin/diyone7 -metadata false -arch AArch64 PodWW PaIL Rfe PodRR Fre
AArch64 MP+poppail+po
{
 int y[2]={0,0};
 0:X1=x; 0:X2=y;
 1:X1=x; 1:X2=y;
}
 P0               | P1          ;
 MOV W0,#1        | LDR W0,[X2] ;
 STR W0,[X1]      | LDR W3,[X1] ;
 MOV W3,#2        |             ;
 SUB W4,W3,#1     |             ;
 STILP W4,W3,[X2] |             ;

exists (1:X0=1 /\ 1:X3=0)

```

And here is one with STLP:

```
herdtools7 % _build/install/default/bin/diyone7 -metadata false -arch AArch64 PodWW PaL Rfe PodRR Fre 
AArch64 MP+poppal+po
{
 int y[2]={0,0};
 0:X1=x; 0:X2=y;
 1:X1=x; 1:X2=y;
}
 P0              | P1          ;
 MOV W0,#1       | LDR W0,[X2] ;
 STR W0,[X1]     | LDR W3,[X1] ;
 MOV W3,#2       |             ;
 SUB W4,W3,#1    |             ;
 STLP W4,W3,[X2] |             ;

exists (1:X0=1 /\ 1:X3=0)

```

RaghavRoy145 commented on lib/AArch64Base.ml:
```diff
@@ -1414,16 +1434,18 @@ type 'k kinstruction =
   | I_STLR of variant * reg * reg
   | I_STXR of variant * st_type * reg * reg * reg
   | I_STXP of variant * st_type * reg * reg * reg * reg
+  | I_STLP of variant * reg * reg * reg
```
I can work on removing these instances and only use LDP/STP variants for these instructions.

HadrienRenaud commented on herd/AArch64ASLSem.ml:
```diff
@@ -866,42 +866,62 @@ module Make (TopConf : AArch64Sig.Config) (V : Value.AArch64ASL) :
+          | `PaIQ -> "ldiappstilp/LDIAPP_32LE_ldiappstilp.opn"
+          | `PaN -> "ldstnapair_offs/LDNP_32_ldstnapair_offs.opn"
+          | `PaA -> "ldiappstilp/LDAP_64_ldiappstilp.opn"
+        and acqrel = pa = `PaIQ && rt != ZR && rt2 != ZR
```
Here you need:

```suggestion
        and acquire = pa = `PaA && rt != ZR && rt2 != ZR
        and acqrel = pa = `PaIQ && rt != ZR && rt2 != ZR
```

RaghavRoy145 replied:
Thanks! I'll try this out

HadrienRenaud commented on herd/AArch64ASLSem.ml:
```diff
@@ -866,42 +866,66 @@ module Make (TopConf : AArch64Sig.Config) (V : Value.AArch64ASL) :
+                "offset" ^= offset;
+                "tagchecked" ^= litb (wback || rn <> SP);
+                (match pa with
+                  | `Pa | `PaN -> with_pos S_Pass
```
Maybe we can inline `with_pos S_Pass` to make it easier to read. This would override the `pass` from `Asllib`, but that's not really a problem:

```suggestion
                  | `Pa | `PaN -> pass
```

HadrienRenaud commented on herd/AArch64ASLSem.ml:
I am happy with the file as it is now.

diaolo01 commented on herd/tests/instructions/AArch64/asl.cfg:
```diff
@@ -61,4 +61,4 @@ nonames L086
 nonames A259
 # Mixed mode non implemented
-nonames M001
+nonames M001
```
What is the change here?

RaghavRoy145 replied:
Thanks for the reviews! I think this is an artefact from my rebase, I will revert this file.

diaolo01 commented on herd/tests/instructions/AArch64/Y006.litmus:
```diff
@@ -0,0 +1,15 @@
+ MOV X3,#1       |                 ;
+ MOV X4,#99      |                 ;
+ STLP X3,X4,[X1] |                 ;
+exists (1:X3=1 /\ 1:X5=0)
```
Would it make sense for this test to require observing both X3 and X4 on P1, so it it clearly testing observation of the STLP pair rather than a standard MP pattern?

RaghavRoy145 replied:
Ah good call, though the behaviour would not change (no witnesses) I should still strengthen the post-condition to include 1:X4=99 so don't permit partial writes

diaolo01 reviewed (commented):
Thanks for this @RaghavRoy145. Generally, this looks good to me.

relokin commented on gen/common/AArch64Arch_gen.ml:
```diff
@@ -240,7 +240,7 @@ type pair_idx = UnspecLoc
   | Plain of capa_opt | Acq of capa_opt | AcqPc of capa_opt | Rel of capa_opt
   | Atomic of atom_rw | Tag | CapaTag | CapaSeal | Pte of atom_pte | Neon of neon_opt
-  | Pair of pair_opt * pair_idx | Instr
+  | Pair of [ld_pair_opt | st_pair_opt] * pair_idx | Instr
```
@ShaleXIONG can you please have a look at the changes (at least to `gen`) of this PR?

ShaleXIONG replied:
On it.

relokin commented on lib/AArch64Parser.mly:
```diff
@@ -566,9 +567,11 @@ zeroopt:
-  { (fun v r1 r2 (r3,idx) -> I_LDP (PaN,v,r1,r2,r3,idx)) }
+  { (fun v r1 r2 (r3,idx) -> I_LDP ((`PaN),v,r1,r2,r3,idx)) }
+| LDAP
+  { (fun v r1 r2 (r3,idx) -> I_LDP ((`PaA),v,r1,r2,r3,idx)) }
```
LDAP doesn't have a post-index variant, so this
`LDAP W0,W1,[X2],#4`
should return an error
and it doesn't have a pre-index variant, so this
`LDAP W0,W1,[X2,#4]!`
should return an error
or any offset other than 0 should also return an error.
`LDAP W0,W1,[X2,#4]`
see https://developer.arm.com/documentation/ddi0602/2026-03/Base-Instructions/LDAP--Load-acquire-pair-of-registers-?lang=en
as opposed to https://developer.arm.com/documentation/ddi0602/2026-03/Base-Instructions/LDP--Load-pair-of-registers-

relokin replied:
one last point, I just realised that LDAP is only defined for reading a pair of 64bit values, so
```
LDP W0,W1,[X2]
```
is not a valid instruction, but
```
LDP X0,X1,[X2]
```
is a valid instruction.

relokin replied:
if you can add these checks in the parser and throw a ParserError when they fail that would be fantastic.

murzinv replied:
> if you can add these checks in the parser and throw a ParserError when they fail that would be fantastic.

I'd suggest using existing infrastructure to reject invalid instructions in user-friendly way - just place checks for invalid combo  under `is_valid` function (`lib/AArch64Base.ml`)

RaghavRoy145 replied:
Thanks for the reviews! with your suggested change @murzinv I am able to reproduce these helpful warnings:

```
Warning: File "./herd/tests/instructions/AArch64/Y006.litmus": Illegal instruction 'LDAP W3,W4,[X0]' (User error)

Warning: File "./herd/tests/instructions/AArch64/Y006.litmus": Illegal instruction 'STLP W3,W4,[X1]' (User error)

Warning: File "./herd/tests/instructions/AArch64/Y006.litmus": Illegal instruction 'STLP X3,X4,[X1,#4]' (User error)
Warning: File "./herd/tests/instructions/AArch64/Y006.litmus": Illegal instruction 'LDAP X3,X4,[X0,#4]' (User error)
```

Though the syntax `LDAP X3,X4,[X0,#0]` or `STLP X3,X4,[X1,#0]` remains allowed

relokin commented on lib/AArch64Parser.mly:
```diff
@@ -589,9 +592,11 @@ ldp_simd_instr:
-  { (fun v r1 r2 (r3,idx) -> I_STP (PaN,v,r1,r2,r3,idx)) }
+  { (fun v r1 r2 (r3,idx) -> I_STP ((`PaN),v,r1,r2,r3,idx)) }
+| STLP
+  { (fun v r1 r2 (r3,idx) -> I_STP ((`PaL),v,r1,r2,r3,idx)) }
```
The same variant restrictions apply for STLP as for LDAP.

relokin reviewed (commented):
Two comments otherwise the herd side looks good to me. The diyone side looks good to me as well but there is also an opportunity to refactor the code there as well, but I will leave that decision to @ShaleXIONG.

relokin commented on lib/AArch64Parser.mly:
```diff
@@ -600,7 +605,7 @@ stp_instr:
           ->
-            I_STP (PaI,v,r1,r2,r3,idx)
+            I_STP ((`PaIL),v,r1,r2,r3,idx)
       | _, _ -> raise Parsing.Parse_error)
```
This is a good example, of the type of checks you need to do.

murzinv replied:
Perhaps both `STILP` and `LDIAPP` deserve moving checks to `is_valid` function, but that for another patch/PR 😄

ShaleXIONG commented on herd/tests/instructions/AArch64/Y006.litmus:
Question in general: should we add more test on the new instruction, `STNP`,`LDNP`, `STILP` and `LDIAPP`?

ShaleXIONG commented:
> @ShaleXIONG can you please review and help with the changes to `gen/`?

Ref: @relokin. I left a few comments. Thanks for @RaghavRoy145 on also extending the `diy` side. if you find it difficult to understand `gen/`, please feel free to ping me and I can help.

relokin commented:
@RaghavRoy145 can you please rebase your PR?

relokin commented on herd/AArch64ASLSem.ml:
```diff
@@ -866,42 +867,66 @@ module Make (TopConf : AArch64Sig.Config) (V : Value.AArch64ASL) :
+                (match pa with
+                  | `Pa | `PaN -> pass
+                  | (`PaA) -> "acquire" ^= litb (rt != ZR && rt2 != ZR)
+                  | (`PaIQ) -> "acqrel" ^= litb (rt != ZR && rt2 != ZR));
```
```suggestion
                  | (`PaA) -> "acquire" ^= litb true
                  | (`PaIQ) -> "acqrel" ^= litb true;
```

RaghavRoy145 replied:
Thanks for this review, I've pushed this change!

relokin commented on herd/AArch64ASLSem.ml:
```diff
@@ -866,42 +867,66 @@ module Make (TopConf : AArch64Sig.Config) (V : Value.AArch64ASL) :
+                (match pa with
+                  | `Pa | `PaN -> pass
+                  | (`PaL) -> "acquire" ^= litb (rt != ZR && rt2 != ZR)
+                  | (`PaIL) -> "acqrel" ^= litb (rt != ZR && rt2 != ZR));
```
```suggestion
                  | (`PaL) -> "acquire" ^= litb true
                  | (`PaIL) -> "acqrel" ^= litb true;
```

relokin reviewed (commented):
Two minor comments, if you can fix these, we can merge this PR.

relokin commented:
Can you also please rebase?

relokin reviewed (approved):
LGTM, thanks Raghav! Will merge this now.
